src/backend/cuda/kernel/memcopy.hpp

```diff
@@ -17,86 +17,248 @@
+template<bool RESHAPE = false>
+void serializeArray(dim_t mdims[4], dim_t mstrides[4], dim_t &mndims,
+                    dim_t ostrides[4], dim_t odims[4] = nullptr) noexcept {
+    if (RESHAPE) assert(odims != nullptr);
+    for (int c = 0; c < mndims - 1; c++) {
+        if (mdims[c] == 1) {
+            // Columns with 1 can always be removed.
+            // strides of the last column are not updated, because:
+            //    - dimension always becomes 1
+            //    - strides are therefor not used
+            for (int i = c; i < mndims - 1; ++i) {
+                mdims[i] = mdims[i + 1];
+                if (RESHAPE) odims[i] = odims[i + 1];
+                mstrides[i] = mstrides[i + 1];
+                ostrides[i] = ostrides[i + 1];
+            }
+            --mndims;
+            mdims[mndims] = 1;
+            if (RESHAPE) odims[mndims] = 1;
+            --c;  // Redo this column, since it is removed now
+        } else if (mdims[c] * mstrides[c] == mstrides[c + 1] &&
+                   mdims[c] * ostrides[c] == ostrides[c + 1]) {
+            // Combine columns, since they are linear
+            // This will increase the dimension of the resulting column,
+            // given more opportunities for kernel optimization
+            mdims[c] *= mdims[c + 1];
+            if (RESHAPE) odims[c] *= odims[c + 1];
+            for (int i = c + 1; i < mndims - 1; ++i) {
+                mdims[i] = mdims[i + 1];
+                if (RESHAPE) odims[i] = odims[i + 1];
+                mstrides[i] = mstrides[i + 1];
+                ostrides[i] = ostrides[i + 1];
+            }
+            --mndims;
+            mdims[mndims] = 1;
+            if (RESHAPE) odims[mndims] = 1;
+            --c;  // Redo this colum, since it is removed now
+        }
+    }
+}
 
-template<typename T>
-void memcopy(Param<T> out, CParam<T> in, const dim_t ndims) {
-    auto memCopy = common::getKernel("cuda::memcopy", {memcopy_cuh_src},
-                                     {TemplateTypename<T>()});
+// To increase the workload inside a kernel, we move a part of the ndims
+// dimension to the last one.  Since this not covered by WG or warp, this is
+// always executed as an internal loop.
+// In case of copy operation, input & output parameters needs to be provided!!
+// main dims: mdims, mstrides, mndims
+// optional dims: odims, ostrides
+// ALL parameters will be updated!!
+void inline increaseWorkload(dim_t elements, dim_t mdims[4], dim_t mstrides[4],
+                             dim_t &mndims, dim_t ostrides[4]) noexcept {
+    if (elements >= 8192 * 2 && mndims != AF_MAX_DIMS && mndims != 0) {
+        // Start only increasing the workload, when all available threads are
+        // occupied.
 
-    dim3 threads(DIMX, DIMY);
+        // list is sorted according to performance improvement
+        // 3x looping is faster than 4x, 2x remains faster than no looping
+        for (const dim_t i : {3, 4, 5, 7, 11, 2}) {
```

Comment:
The 8192 should correspond to the reasoning in dispatch.hpp (calculation of minThreads).

I will add those constant to dispatch.hpp (since it is already included)
The constants will have the corresponding explanation as well.